contracts/random/RandomSeedProvider.sol

```diff
@@ -0,0 +1,297 @@
+     * @notice Called during contract upgrade.
+     * @dev This function will be overridden in future versions of this contract.
+     */
+    function upgrade() external virtual {
+        // Revert in the following situations:
+        // - The function is called on the existing version.
+        // - This version of code is mistakenly deploy, for an upgrade from V2 to V3.
+        //   That is, we mistakenly attempt to downgrade the contract.
+        revert CanNotUpgradeFrom(version, VERSION0);
+    }
+
+    /**
+     * @notice Change the off-chain random source.
+     * @dev Only RANDOM_ADMIN_ROLE can do this.
+     * @param _offchainRandomSource Address of contract that is an off-chain random source.
+     */
+    function setOffchainRandomSource(address _offchainRandomSource) external onlyRole(RANDOM_ADMIN_ROLE) {
+        randomSource = _offchainRandomSource;
+        emit OffchainRandomSourceSet(_offchainRandomSource);
+    }
+
+    /**
+     * @notice Call this when the blockchain supports the PREVRANDAO opcode.
+     * @dev Only RANDOM_ADMIN_ROLE can do this.
+     */
+    function setRanDaoAvailable() external onlyRole(RANDOM_ADMIN_ROLE) {
+        ranDaoAvailable = true;
+        emit RanDaoEnabled();
+    }
+
+    /**
+     * @notice Add a consumer that can use off-chain supplied random.
+     * @dev Only RANDOM_ADMIN_ROLE can do this.
+     * @param _consumer Game contract that inherits from RandomValues.sol that is authorised to use off-chain random.
+     */
+    function addOffchainRandomConsumer(address _consumer) external onlyRole(RANDOM_ADMIN_ROLE) {
+        approvedForOffchainRandom[_consumer] = true;
+        emit OffchainRandomConsumerAdded(_consumer);
+    }
+
+    /**
+     * @notice Remove a consumer that can use off-chain supplied random.
+     * @dev Only RANDOM_ADMIN_ROLE can do this.
+     * @param _consumer Game contract that inherits from RandomValues.sol that is no longer authorised to use off-chain random.
+     */
+    function removeOffchainRandomConsumer(address _consumer) external onlyRole(RANDOM_ADMIN_ROLE) {
+        approvedForOffchainRandom[_consumer] = false;
+        emit OffchainRandomConsumerRemoved(_consumer);
+    }
+
+    /**
+     * @notice Request the index number to track when a random number will be produced.
+     * @dev Note that the same _randomFulfillmentIndex will be returned to multiple games and even within
+     * @dev the one game. Games must personalise this value to their own game, the particular game player,
+     * @dev and to the game player's request.
+     * @return _randomFulfillmentIndex The index for the game contract to present to fetch the next random value.
+     * @return _randomSource Indicates that an on-chain source was used, or is the address of an off-chain source.
+     */
+    function requestRandomSeed() external returns (uint256 _randomFulfillmentIndex, address _randomSource) {
+        if (randomSource == ONCHAIN || !approvedForOffchainRandom[msg.sender]) {
```

Comment:
if a game relies on offchain randomness but is not approved, it should error instead of giving them onchain randomness without telling them

Reply:
This API does something a little different: the game doesn't specify where the random values should come from. It asks from a random number. A game contract could check the result of the call, and then fail if the source was unacceptable. 

I did consider having an API where the source could be specified. However, this adds complexity. It means that the game studio would need to be able to specify the source, and change as we introduced off-chain sources. Doing it this way should allow the game studio to use random numbers without changing their code.